File: string-calculator/strign-calculator-java/src/main/java/pl/frognar/_02/StringCalculator.java
package pl.frognar._02;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class StringCalculator {
    public int add(String numbers) {
        if (numbers.isEmpty()) {
            return 0;
        }
        var numbersList = convertToInts(splitAndFilterNumbers(numbers));
        var negatives = numbersList.stream().filter(n -> n < 0).toList();
        if (!negatives.isEmpty()) {
            String negativesString = negatives
                    .stream()
                    .map(String::valueOf)
                    .collect(Collectors.joining(","));
            throw new NegativeNumbersNotAllowedException(
                    "negatives not allowed: %s".formatted(negativesString));
        }

        return numbersList.stream().reduce(0, Integer::sum);
    }

    Stream<String> splitAndFilterNumbers(String numbers) {
        return Arrays.stream(numbers.replaceAll("[^0-9-]", ",")
                .split(","))
                .filter(s -> !s.isEmpty());
    }

    List<Integer> convertToInts(Stream<String> numbers) {
        return numbers
                .map(Integer::parseInt)
                .toList();
    }
}

File: string-calculator/string-calculator-py/03/string_calculator.py
from re import split
from typing import List


class StringCalculator:
    @classmethod
    def add(cls, numbers: str) -> int:
        if not numbers:
            return 0
        return sum(cls.split_and_convert(numbers))

    @staticmethod
    def split_and_convert(numbers: str) -> List[int]:
        default_delimiter = '[,\n]'
        return [int(n) for n in split(default_delimiter, numbers)]